File: python/tests/test_anya.py
"""Anya — the TRUE Euclidean any-angle optimum (turns at grid corners, not cell
centres): equals the corner-visibility optimum, never exceeds Visibility A*, and
the open-space / no-path / parameter-validation contracts."""

from __future__ import annotations

import heapq
import io
import json
import math
import random
from pathlib import Path

import pytest
from conftest import config, grid_from, open_grid, write_config

from navigation.core.params import ParamError, ParamSet
from navigation.core.trace import TraceRecorder
from navigation.core.types import Cell
from navigation.global_planning.search.anya import Anya
from navigation.global_planning.search.theta_star import ThetaStar
from navigation.global_planning.search.visibility_astar import VisibilityAStarPlanner

# --- independent ground truth: corner visibility graph ----------------------
# Turning points of a shortest Euclidean path in a polygonal (blocked-cell) domain
# lie only at convex obstacle corners (grid vertices), so the optimum is a Dijkstra
# over {start, goal, all obstacle corners} weighted by straight-line length. The
# line-of-sight oracle below is computed by a DIFFERENT method than the planner
# (analytic Liang-Barsky clipping against open cell squares, not midpoint sampling),
# so agreement is a real correctness check, not a restatement of the planner.

_Pt = tuple[float, float]


def _reachable(grid: object, start: Cell) -> set[Cell]:
    seen: set[Cell] = {start}
    stack = [start]
    while stack:
        cur = stack.pop()
        for nb, _c in grid.neighbors(cur):  # type: ignore[attr-defined]
            if nb not in seen:
                seen.add(nb)
                stack.append(nb)
    return seen


def _free(cells: set[Cell], cx: int, cy: int) -> bool:
    return (cy, cx) in cells


def _open_slab(p: float, d: float, lo: float, hi: float) -> tuple[float, float] | None:
    # t-interval where p + t*d lies strictly inside the open interval (lo, hi); None
    # if empty. An axis-parallel component that stays exactly on a boundary (d == 0
    # and p == lo/hi) yields None, so a segment grazing a cell edge is NOT counted as
    # crossing that cell's interior.
    if d == 0.0:
        return (-math.inf, math.inf) if lo < p < hi else None
    t0, t1 = (lo - p) / d, (hi - p) / d
    return (t0, t1) if t0 <= t1 else (t1, t0)


def _seg_clear(cells: set[Cell], p: _Pt, q: _Pt, eps: float = 1e-9) -> bool:
    """Exact grazing-aware any-angle line of sight, computed independently of the
    planner. A segment is valid iff (1) it enters no blocked cell interior (analytic
    Liang-Barsky clip against each cell's OPEN square — a positive-length overlap is a
    real interior crossing; touching an edge is not), (2) it does not run along a grid
    edge whose two sides are both blocked, and (3) it does not corner-cut through a
    pinch of two diagonally blocked cells. Edge-grazing an obstacle corner is allowed,
    matching the true continuous Euclidean model (Harabor et al. 2016)."""
    px, py = p
    qx, qy = q
    if p == q:
        return True
    dx, dy = qx - px, qy - py
    xlo, xhi = sorted((px, qx))
    ylo, yhi = sorted((py, qy))
    # (1) interior crossing
    for cy in range(math.floor(ylo), math.ceil(yhi)):
        for cx in range(math.floor(xlo), math.ceil(xhi)):
            if _free(cells, cx, cy):
                continue
            sx = _open_slab(px, dx, cx, cx + 1)
            sy = _open_slab(py, dy, cy, cy + 1)
            if sx is None or sy is None:
                continue
            if min(sx[1], sy[1], 1.0) - max(sx[0], sy[0], 0.0) > 1e-9:
                return False
    # (2) both-sides-blocked grid edge (axis segment exactly on an integer line)
    if dx == 0.0 and abs(px - round(px)) < eps:
        xi = round(px)
        for row in range(math.floor(ylo), math.ceil(yhi)):
            if ylo < row + 1 and yhi > row and not _free(cells, xi - 1, row) \
                    and not _free(cells, xi, row):
                return False
    if dy == 0.0 and abs(py - round(py)) < eps:
        yi = round(py)
        for col in range(math.floor(xlo), math.ceil(xhi)):
            if xlo < col + 1 and xhi > col and not _free(cells, col, yi - 1) \
                    and not _free(cells, col, yi):
                return False
    # (3) pinch (corner-cutting) at an interior lattice corner of a diagonal segment
    if dx != 0.0 and dy != 0.0:
        for ix in range(math.ceil(xlo), math.floor(xhi) + 1):
            t = (ix - px) / dx
            if t <= eps or t >= 1.0 - eps:
                continue
            y = py + t * dy
            if abs(y - round(y)) < eps:
                iy = round(y)
                if (dx > 0.0) == (dy > 0.0):
                    if not _free(cells, ix - 1, iy) and not _free(cells, ix, iy - 1):
                        return False
                elif not _free(cells, ix - 1, iy - 1) and not _free(cells, ix, iy):
                    return False
    return True


def _corner_optimum(grid: object, start: Cell, goal: Cell) -> float:
    cells = _reachable(grid, start)  # same domain the planner observes
    if goal not in cells:
        return math.inf
    rows = [r for r, _c in cells]
    cols = [c for _r, c in cells]
    verts: list[_Pt] = []
    for y in range(min(rows), max(rows) + 2):
        for x in range(min(cols), max(cols) + 2):
            around = [not _free(cells, x - 1, y - 1), not _free(cells, x, y - 1),
                      not _free(cells, x - 1, y), not _free(cells, x, y)]
            if any(around) and not all(around):
                verts.append((float(x), float(y)))
    s: _Pt = (start[1] + 0.5, start[0] + 0.5)
    g: _Pt = (goal[1] + 0.5, goal[0] + 0.5)
    nodes = list(dict.fromkeys([s, g, *verts]))
    dist = {s: 0.0}
    pq: list[tuple[float, int]] = [(0.0, 0)]
    settled: set[int] = set()
    while pq:
        d, ui = heapq.heappop(pq)
        u = nodes[ui]
        if ui in settled:
            continue
        settled.add(ui)
        if u == g:
            return d
        for vi, v in enumerate(nodes):
            if vi in settled or v == u:
                continue
            if _seg_clear(cells, u, v):
                nd = d + math.dist(u, v)
                if nd < dist.get(v, math.inf):
                    dist[v] = nd
                    heapq.heappush(pq, (nd, vi))
    return math.inf


# (a) OPTIMALITY: corner turning beats cell-centre ---------------------------

def test_anya_turns_at_corner_and_beats_cell_centre() -> None:
    # A vertical bar hides the goal; the shortest route grazes the bar's top-left
    # grid corner (2,2). Turning THERE (not at any cell centre) gives the true
    # Euclidean optimum 2*sqrt(8.5); Visibility A* / Theta*, pinned to cell centres,
    # must be strictly longer. This is Anya's defining property.
    grid = grid_from([".....", ".....", "..#..", "..#..", "....."])
    start, goal = (4, 0), (0, 4)
    res = Anya(config("anya")).plan(grid, start, goal)
    assert res.success
    assert res.path[0] == start and res.path[-1] == goal
    # exact corner-turning optimum: start->corner(2,2)->goal, each leg sqrt(8.5)
    assert res.cost == pytest.approx(2.0 * math.sqrt(8.5))
    # equals the independent corner-visibility optimum
    assert res.cost == pytest.approx(_corner_optimum(grid, start, goal))
    # strictly shorter than the cell-centre any-angle planners
    vis = VisibilityAStarPlanner(config("visibility_astar")).plan(grid, start, goal)
    theta = ThetaStar(config("theta_star")).plan(grid, start, goal)
    assert res.cost < vis.cost - 1e-9
    assert res.cost < theta.cost - 1e-9


def test_anya_matches_corner_optimum_on_asymmetric_map() -> None:
    # A second, asymmetric instance: guards against a projection correct only on the
    # symmetric bend above. Anya must still equal the corner-visibility optimum and
    # never exceed Visibility A*.
    grid = grid_from([
        "........",
        "...##...",
        "...##...",
        ".....#..",
        "..#.....",
        "........",
    ])
    start, goal = (5, 0), (0, 7)
    res = Anya(config("anya")).plan(grid, start, goal)
    assert res.success
    assert res.cost == pytest.approx(_corner_optimum(grid, start, goal))
    vis = VisibilityAStarPlanner(config("visibility_astar")).plan(grid, start, goal)
    assert res.cost <= vis.cost + 1e-9


# (a') EDGE-GRAZING optimum: a taut leg that hugs an obstacle edge --------------
# These are the instances where snapping the on-grid-line line-of-sight sample to one
# side (floor) forbids a legal edge-grazing leg and inflates the cost. The optimum is
# only reachable if a segment that runs exactly along a blocked cell's boundary (but
# through no interior) is admitted.

def test_anya_optimum_hugs_blocked_cell_edge() -> None:
    # (3,1)->(2,3)->(3,3): the middle leg grazes the top edge of blocked cell (3,2)
    # without entering its interior. Optimum 1 + sqrt(2); a floor-snapped LOS instead
    # detours over (1,1) at cost 2*sqrt(8)/... > 3.7.
    grid = grid_from([".##.", "..#.", "....", "..#."])
    start, goal = (3, 1), (3, 3)
    res = Anya(config("anya")).plan(grid, start, goal)
    assert res.success
    assert res.cost == pytest.approx(1.0 + math.sqrt(2.0))
    assert res.cost == pytest.approx(_corner_optimum(grid, start, goal))


def test_anya_never_exceeds_visibility_astar_on_corner_turn() -> None:
    # A corner-turning instance where the floor-snapped LOS made Anya STRICTLY worse
    # than Visibility A* (cell-centre). The true any-angle optimum must be <= it.
    grid = grid_from([".......", "#.##...", "##...#.", ".....#.", "....#.."])
    start, goal = (2, 6), (2, 2)
    res = Anya(config("anya")).plan(grid, start, goal)
    assert res.success
    assert res.cost == pytest.approx(_corner_optimum(grid, start, goal))
    vis = VisibilityAStarPlanner(config("visibility_astar")).plan(grid, start, goal)
    assert res.cost <= vis.cost + 1e-9


# (a'') PROPERTY: Anya == exact grazing-aware corner optimum on random maps ------

@pytest.mark.parametrize("seed", range(600))
def test_anya_matches_exact_oracle_on_random_grids(seed: int) -> None:
    # Anya's cost must equal the independent, grazing-aware corner-visibility optimum
    # on every solvable random instance. The oracle uses analytic Liang-Barsky clips
    # (not the planner's sampled LOS), so this pins true continuous Euclidean
    # optimality, not a self-consistency check.
    rng = random.Random(seed)
    rows = rng.randint(3, 6)
    cols = rng.randint(3, 6)
    density = rng.uniform(0.12, 0.32)
    ascii_rows = [
        "".join("#" if rng.random() < density else "." for _ in range(cols))
        for _ in range(rows)
    ]
    free_cells = [(r, c) for r in range(rows) for c in range(cols) if ascii_rows[r][c] == "."]
    if len(free_cells) < 2:
        pytest.skip("degenerate map")
    start = rng.choice(free_cells)
    goal = rng.choice(free_cells)
    if start == goal:
        pytest.skip("start == goal")
    grid = grid_from(ascii_rows)
    if goal not in _reachable(grid, start):
        pytest.skip("goal unreachable")
    res = Anya(config("anya")).plan(grid, start, goal)
    assert res.success
    assert res.cost == pytest.approx(_corner_optimum(grid, start, goal), abs=1e-9)


# (b) open space: straight line ----------------------------------------------

def test_anya_returns_straight_line_when_visible() -> None:
    grid = open_grid(4, 4)
    start, goal = (3, 0), (0, 3)
    res = Anya(config("anya")).plan(grid, start, goal)
    assert res.success
    assert res.path[0] == start and res.path[-1] == goal
    assert res.cost == pytest.approx(math.hypot(3, 3))  # single unobstructed segment


# trace surfaces the (root, interval) node ------------------------------------

def test_anya_emits_root_interval_data() -> None:
    buf = io.StringIO()
    Anya(config("anya")).plan(grid_from([".....", ".....", "..#..", "..#..", "....."]),
                              (4, 0), (0, 4), TraceRecorder(buf))
    intervals = [
        json.loads(line)["data"]
        for line in buf.getvalue().splitlines()
        if json.loads(line)["event"] == "edge_added"
    ]
    assert intervals, "any-angle edges must be tagged with their (root, interval) node"
    for d in intervals:
        assert d.keys() == {"row", "col_lo", "col_hi"}
        assert d["col_lo"] <= d["col_hi"] + 1e-9


# (c) no-path case -----------------------------------------------------------

def test_anya_no_path_when_walled_off() -> None:
    grid = grid_from(["..#..", "..#..", "..#..", "..#..", "..#.."])
    res = Anya(config("anya")).plan(grid, (0, 0), (0, 4))
    assert not res.success
    assert res.path == []
    assert res.cost == 0.0


# (d) parameter validation failure -------------------------------------------

def test_anya_rejects_out_of_range_epsilon(tmp_path: Path) -> None:
    # anya declares vertex_epsilon in [1e-12, 1e-3]; an above-max default must fail
    # at load time so an invalid tolerance can never reach plan().
    cfg = write_config(
        tmp_path / "bad_anya.yaml",
        "anya",
        [{"name": "vertex_epsilon", "type": "float", "default": 1.0,
          "min": 1.0e-12, "max": 1.0e-3, "description": "above max"}],
    )
    with pytest.raises(ParamError):
        ParamSet.from_yaml(cfg)

File: cpp/tests/test_anya.cpp
#include <algorithm>
#include <cmath>
#include <limits>
#include <queue>
#include <random>
#include <set>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

#include <gtest/gtest.h>

#include "navigation/global_planning/search/anya.hpp"
#include "navigation/global_planning/search/theta_star.hpp"
#include "navigation/global_planning/search/visibility_astar.hpp"
#include "test_util.hpp"

using namespace navigation;
using core::Cell;

namespace {

core::ParamSet cfg(const std::string& algo) {
  return core::ParamSet::from_yaml(test::repo_path("configs/global_planning/" + algo + ".yaml"));
}

using Pt = std::pair<double, double>;

// Independent ground truth: turning points of a shortest Euclidean path in a
// blocked-cell domain lie only at convex obstacle corners (grid vertices), so the
// optimum is an all-pairs corner-visibility Dijkstra with a pinch-aware corner LOS
// over the same reachable component the planner observes. A different computation
// than Anya's interval sweep, so agreement is a real correctness check.

bool cfree(const std::unordered_set<Cell>& free, int cx, int cy) {
  return free.count(Cell{cy, cx}) > 0;
}

// t-interval where p + t*d lies strictly inside the open interval (lo, hi); ok=false
// if empty. A component staying exactly on a boundary (d==0, p==lo/hi) is empty, so
// grazing a cell edge is not counted as crossing that cell's interior.
struct Slab {
  bool ok;
  double t0, t1;
};
Slab open_slab(double p, double d, double lo, double hi) {
  if (d == 0.0) return (lo < p && p < hi) ? Slab{true, -1e300, 1e300} : Slab{false, 0.0, 0.0};
  double a = (lo - p) / d, b = (hi - p) / d;
  return a <= b ? Slab{true, a, b} : Slab{true, b, a};
}

// Exact grazing-aware any-angle line of sight, computed independently of the planner:
// (1) analytic Liang-Barsky clip against each blocked cell's OPEN square (positive
// overlap = real interior crossing; touching an edge is not), (2) no travel along a
// grid edge with both sides blocked, (3) no corner-cut through a diagonal blocked
// pinch. Edge-grazing an obstacle corner is allowed (Harabor et al. 2016).
bool seg_clear(const std::unordered_set<Cell>& free, const Pt& p, const Pt& q, double eps = 1e-9) {
  if (p == q) return true;
  double px = p.first, py = p.second, qx = q.first, qy = q.second;
  double dx = qx - px, dy = qy - py;
  double xlo = std::min(px, qx), xhi = std::max(px, qx);
  double ylo = std::min(py, qy), yhi = std::max(py, qy);
  for (int cy = static_cast<int>(std::floor(ylo)); cy < static_cast<int>(std::ceil(yhi)); ++cy) {
    for (int cx = static_cast<int>(std::floor(xlo)); cx < static_cast<int>(std::ceil(xhi)); ++cx) {
      if (cfree(free, cx, cy)) continue;
      Slab sx = open_slab(px, dx, cx, cx + 1), sy = open_slab(py, dy, cy, cy + 1);
      if (!sx.ok || !sy.ok) continue;
      double t0 = std::max({sx.t0, sy.t0, 0.0}), t1 = std::min({sx.t1, sy.t1, 1.0});
      if (t1 - t0 > 1e-9) return false;
    }
  }
  if (dx == 0.0 && std::abs(px - std::round(px)) < eps) {
    int xi = static_cast<int>(std::round(px));
    for (int row = static_cast<int>(std::floor(ylo)); row < static_cast<int>(std::ceil(yhi)); ++row)
      if (ylo < row + 1 && yhi > row && !cfree(free, xi - 1, row) && !cfree(free, xi, row))
        return false;
  }
  if (dy == 0.0 && std::abs(py - std::round(py)) < eps) {
    int yi = static_cast<int>(std::round(py));
    for (int col = static_cast<int>(std::floor(xlo)); col < static_cast<int>(std::ceil(xhi)); ++col)
      if (xlo < col + 1 && xhi > col && !cfree(free, col, yi - 1) && !cfree(free, col, yi))
        return false;
  }
  if (dx != 0.0 && dy != 0.0) {
    for (int ix = static_cast<int>(std::ceil(xlo)); ix <= static_cast<int>(std::floor(xhi)); ++ix) {
      double t = (ix - px) / dx;
      if (t <= eps || t >= 1.0 - eps) continue;
      double y = py + t * dy;
      if (std::abs(y - std::round(y)) < eps) {
        int iy = static_cast<int>(std::round(y));
        if ((dx > 0.0) == (dy > 0.0)) {
          if (!cfree(free, ix - 1, iy) && !cfree(free, ix, iy - 1)) return false;
        } else if (!cfree(free, ix - 1, iy - 1) && !cfree(free, ix, iy)) {
          return false;
        }
      }
    }
  }
  return true;
}

double corner_optimum(maps::OccupancyGrid2D& grid, const Cell& start, const Cell& goal) {
  std::unordered_set<Cell> free{start};
  std::vector<Cell> st{start};
  while (!st.empty()) {
    Cell c = st.back();
    st.pop_back();
    for (const auto& [nb, w] : grid.neighbors(c)) {
      (void)w;
      if (free.insert(nb).second) st.push_back(nb);
    }
  }
  if (!free.count(goal)) return std::numeric_limits<double>::infinity();
  int r0 = start.row, r1 = start.row, c0 = start.col, c1 = start.col;
  for (const Cell& c : free) {
    r0 = std::min(r0, c.row);
    r1 = std::max(r1, c.row);
    c0 = std::min(c0, c.col);
    c1 = std::max(c1, c.col);
  }
  std::vector<Pt> nodes{{start.col + 0.5, start.row + 0.5}, {goal.col + 0.5, goal.row + 0.5}};
  for (int y = r0; y <= r1 + 1; ++y) {
    for (int x = c0; x <= c1 + 1; ++x) {
      int blocked = 0;
      for (auto [cx, cy] : {std::pair{x - 1, y - 1}, {x, y - 1}, {x - 1, y}, {x, y}})
        blocked += cfree(free, cx, cy) ? 0 : 1;
      if (blocked > 0 && blocked < 4) nodes.push_back({static_cast<double>(x), static_cast<double>(y)});
    }
  }
  int n = static_cast<int>(nodes.size());
  std::vector<double> dist(n, std::numeric_limits<double>::infinity());
  std::vector<bool> settled(n, false);
  dist[0] = 0.0;
  using QI = std::pair<double, int>;
  std::priority_queue<QI, std::vector<QI>, std::greater<QI>> pq;
  pq.push({0.0, 0});
  while (!pq.empty()) {
    auto [d, u] = pq.top();
    pq.pop();
    if (settled[u]) continue;
    settled[u] = true;
    if (u == 1) return d;
    for (int v = 0; v < n; ++v) {
      if (settled[v] || v == u) continue;
      if (seg_clear(free, nodes[u], nodes[v])) {
        double nd = d + std::hypot(nodes[u].first - nodes[v].first,
                                   nodes[u].second - nodes[v].second);
        if (nd < dist[v]) {
          dist[v] = nd;
          pq.push({nd, v});
        }
      }
    }
  }
  return std::numeric_limits<double>::infinity();
}

}  // namespace

// (a) OPTIMALITY: turning at a grid corner beats every cell-centre planner --------

// A vertical bar hides the goal; the shortest route grazes the bar's top-left grid
// corner (2,2). Turning THERE (not at any cell centre) yields the true Euclidean
// optimum 2*sqrt(8.5); Visibility A* / Theta*, pinned to cell centres, are strictly
// longer. This is Anya's defining property.
TEST(Anya, TurnsAtCornerAndBeatsCellCentre) {
  auto g = test::make_grid({".....", ".....", "..#..", "..#..", "....."});
  Cell start{4, 0}, goal{0, 4};
  auto ra = global_planning::AnyaPlanner(cfg("anya")).plan(g, start, goal, nullptr);
  ASSERT_TRUE(ra.success);
  EXPECT_EQ(ra.path.front(), start);
  EXPECT_EQ(ra.path.back(), goal);
  EXPECT_NEAR(ra.cost, 2.0 * std::sqrt(8.5), 1e-9);
  EXPECT_NEAR(ra.cost, corner_optimum(g, start, goal), 1e-9);
  global_planning::VisibilityAStarPlanner vis(cfg("visibility_astar"));
  global_planning::ThetaStarPlanner theta(cfg("theta_star"));
  EXPECT_LT(ra.cost, vis.plan(g, start, goal, nullptr).cost - 1e-9);
  EXPECT_LT(ra.cost, theta.plan(g, start, goal, nullptr).cost - 1e-9);
}

// A second, asymmetric instance guards against a projection correct only on the
// symmetric bend above.
TEST(Anya, MatchesCornerOptimumOnAsymmetricMap) {
  auto g = test::make_grid({
      "........", "...##...", "...##...", ".....#..", "..#.....", "........",
  });
  Cell start{5, 0}, goal{0, 7};
  auto ra = global_planning::AnyaPlanner(cfg("anya")).plan(g, start, goal, nullptr);
  ASSERT_TRUE(ra.success);
  EXPECT_NEAR(ra.cost, corner_optimum(g, start, goal), 1e-9);
  global_planning::VisibilityAStarPlanner vis(cfg("visibility_astar"));
  EXPECT_LE(ra.cost, vis.plan(g, start, goal, nullptr).cost + 1e-9);
}

// (a') EDGE-GRAZING optimum: a taut leg hugs an obstacle edge without entering its
// interior. Snapping the on-grid-line line-of-sight sample to one side (floor) would
// forbid such a leg and inflate the cost. Optimum 1 + sqrt(2) here.
TEST(Anya, OptimumHugsBlockedCellEdge) {
  auto g = test::make_grid({".##.", "..#.", "....", "..#."});
  Cell start{3, 1}, goal{3, 3};
  auto ra = global_planning::AnyaPlanner(cfg("anya")).plan(g, start, goal, nullptr);
  ASSERT_TRUE(ra.success);
  EXPECT_NEAR(ra.cost, 1.0 + std::sqrt(2.0), 1e-9);
  EXPECT_NEAR(ra.cost, corner_optimum(g, start, goal), 1e-9);
}

// A corner-turning instance where floor-snapped LOS made Anya STRICTLY worse than
// Visibility A*; the true any-angle optimum must never exceed it.
TEST(Anya, NeverExceedsVisibilityAStarOnCornerTurn) {
  auto g = test::make_grid({".......", "#.##...", "##...#.", ".....#.", "....#.."});
  Cell start{2, 6}, goal{2, 2};
  auto ra = global_planning::AnyaPlanner(cfg("anya")).plan(g, start, goal, nullptr);
  ASSERT_TRUE(ra.success);
  EXPECT_NEAR(ra.cost, corner_optimum(g, start, goal), 1e-9);
  global_planning::VisibilityAStarPlanner vis(cfg("visibility_astar"));
  EXPECT_LE(ra.cost, vis.plan(g, start, goal, nullptr).cost + 1e-9);
}

// (a'') PROPERTY: Anya == exact grazing-aware corner optimum on random maps. The
// oracle uses analytic Liang-Barsky clips (a different computation than the planner's
// sampled sweep), so agreement pins true continuous Euclidean optimality.
TEST(Anya, MatchesExactOracleOnRandomGrids) {
  int checked = 0;
  for (unsigned seed = 0; seed < 600; ++seed) {
    std::mt19937 rng(seed);
    int rows = std::uniform_int_distribution<int>(3, 6)(rng);
    int cols = std::uniform_int_distribution<int>(3, 6)(rng);
    double density = std::uniform_real_distribution<double>(0.12, 0.32)(rng);
    std::vector<std::string> ascii(rows, std::string(cols, '.'));
    std::vector<Cell> frees;
    for (int r = 0; r < rows; ++r)
      for (int c = 0; c < cols; ++c) {
        if (std::uniform_real_distribution<double>(0.0, 1.0)(rng) < density)
          ascii[r][c] = '#';
        else
          frees.push_back(Cell{r, c});
      }
    if (frees.size() < 2) continue;
    Cell start = frees[std::uniform_int_distribution<size_t>(0, frees.size() - 1)(rng)];
    Cell goal = frees[std::uniform_int_distribution<size_t>(0, frees.size() - 1)(rng)];
    if (start == goal) continue;
    auto g = test::make_grid(ascii);
    double opt = corner_optimum(g, start, goal);
    if (std::isinf(opt)) continue;  // goal unreachable — no path to compare
    auto ra = global_planning::AnyaPlanner(cfg("anya")).plan(g, start, goal, nullptr);
    ASSERT_TRUE(ra.success) << "seed " << seed;
    EXPECT_NEAR(ra.cost, opt, 1e-9) << "seed " << seed;
    ++checked;
  }
  EXPECT_GT(checked, 200);  // sanity: a meaningful number of solvable instances ran
}

// (b) open space: a single straight segment ---------------------------------------

TEST(Anya, ReturnsStraightLineWhenVisible) {
  auto g = test::make_grid({"....", "....", "....", "...."});
  Cell start{3, 0}, goal{0, 3};
  auto ra = global_planning::AnyaPlanner(cfg("anya")).plan(g, start, goal, nullptr);
  ASSERT_TRUE(ra.success);
  EXPECT_EQ(ra.path.front(), start);
  EXPECT_EQ(ra.path.back(), goal);
  EXPECT_NEAR(ra.cost, std::sqrt(18.0), 1e-9);
}

// (c) no-path case ----------------------------------------------------------------

TEST(Anya, NoPathWhenWalledOff) {
  auto g = test::make_grid({"..#..", "..#..", "..#..", "..#..", "..#.."});
  auto ra = global_planning::AnyaPlanner(cfg("anya")).plan(g, Cell{0, 0}, Cell{0, 4}, nullptr);
  EXPECT_FALSE(ra.success);
  EXPECT_TRUE(ra.path.empty());
  EXPECT_EQ(ra.cost, 0.0);
}

// (d) param validation failure ----------------------------------------------------

TEST(Anya, BadParamThrows) {
  std::string bad = test::write_temp(
      "anya.yaml",
      "algorithm: anya\ncategory: global_planning\nparams:\n"
      "  - name: vertex_epsilon\n    type: float\n    default: 1.0\n    min: 1.0e-12\n"
      "    max: 1.0e-3\n    description: above max\n");
  EXPECT_THROW(core::ParamSet::from_yaml(bad), std::runtime_error);
}
